rfcs/2021-10-29-8621-framing-and-codecs-sinks.md

```diff
@@ -0,0 +1,126 @@
+
+Another related current work is `StandardEncoding` which has been introduced in [#9215](https://github.com/vectordotdev/vector/pull/9215). It provides a consistent set of codec implementations aims and therefore aligns with our goals. However, it does not separate the reshaping responsibilities.
+
+## Scope
+
+### In scope
+
+- Common support structures for encoding/framing
+- Set of commonly used encoding (text, JSON, ...) and framing (newline-delimited, octet-counting, ...) methods
+- Strategy for integrating encoding/framing with sinks
+
+### Out of scope
+
+- Integration with schema system (automatic transformation of event fields depending)
+
+## Pain
+
+Internally, we don't share implementations for encoding strategies that are very commonly used (e.g. JSON, newline-delimited). This leads to duplicated effort and missing guidance for handling encoding in sinks since no canonical implementation exists.
+
+Externally, users are restricted by the available custom set of encodings implemented by the sink. There does not exist any mental model for consistent encoding/framing options for event data that applies to all sinks.
+
+## Proposal
+
+### User Experience
+
+Users should be able to set `encoding` and `framing` options on sinks, analogously to the `decoding` and `framing` options on sources. These options uniformly control how the event _payload_ is encoded. This distinction is important, as encoding for the sink specific _protocol_ and the event _payload_ are separate concerns. The payload should still be encoded according to the sink's protocol, and the sink should provide additional options if there are multiple protocols to choose from, e.g. under a `protocol` key.
+
+The fields containing event transformations in sinks on the current `encoding` options (`schema`, `only_fields`, `except_fields`, `timestamp_format`) should be moved to a dedicated option, e.g. `schema` or `transform`. Thus, this would introduce a breaking change in configurations. However, migration would be relatively straight-forward by nesting these options under the new key.
+
+### Implementation
+
+The following common config structures will be provided, analogously to the decoding/framing configs:
+
+```rust
+pub trait SerializerConfig: Debug + DynClone + Send + Sync {
+    /// Builds a serializer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedSerializer>;
+}
+```
+
+```rust
+pub trait FramingConfig: Debug + DynClone + Send + Sync {
+    /// Builds a framer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedFramer>;
+}
+```
+
+These can be build to form an `Encoder`:
+
+```rust
+/// An encoder that can encode structured events to byte messages.
+pub struct Encoder {
+    serializer: BoxedSerializer,
+    framer: BoxedFramer,
+}
+```
```

Comment:
One thing I'd like to see here is if we can actually make `SerializerConfig`/`FramingConfig` provide their outputs generically.  For example:

```rust
pub trait SerializerConfig {
    type Serializer;
    
    fn build(&self) -> crate::Result<Self::Serializer>;
}

pub trait FramingConfig {
    type Framer;
    
    fn build(&self) -> crate::Result<Self::Framer>;
}

pub struct Encoder<S, F> {
    serializer: S,
    framer: F,
}
```

This is sort of an implementation detail, but we should avoid boxing individual components where possible, and prefer to box at a higher level.  It might even be possible to entirely avoid boxing.  I'm bringing this up here on the off-chance that not using boxed inner fields would meaningfully change how we have to approach the implementation.

Since codecs touch every single event entering and exiting Vector, we should take care to avoid boxing where possible.

Reply:
Hmm, this is a very interesting point.

Since the source side is structured very similarly, I tried to apply your suggestion to the `FramingConfig` and `ParserConfig`.

It fails to compile since `dyn_clone::clone_trait_object` needs to specify the associated type. Cloning a decoder/encoder is a rather common operation, it happens for each message or stream to reset the codec state.

I'm not sure if it's possible to circumvent this without having a `Box` on the inner type, but I'd be happy for any suggestions.